### PR TITLE
add Video feature and kernels

### PR DESCRIPTION
**Pull request by pmeier**

### CODE REVIEW

File: torchvision/prototype/features/_video.py
Comment (pmeier) [inline torchvision/prototype/features/_video.py:64]: This is new for videos. Is that the correct name? cc @bjuncek @jdsgomes

File: torchvision/prototype/transforms/functional/_augment.py
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_augment.py:20]: The new video kernels are trivial for now, since they only wrap the image kernels. This works as long as the image kernel supports arbitrary batch sizes, since for all the current transformations videos can be regarded as a batch of images. 

However, there are some kernels that currently don't support arbitrary batch sizes:

https://github.com/pytorch/vision/blob/9cece405db3a8301463b219d9e0885142837ae4f/test/prototype_transforms_kernel_infos.py#L906-L907

We need to fix them in order for the video kernel to work properly.

In the future, if we have transforms that work on the temporal dimension, i.e. `video.shape[-4]`, we actually need to implement explicit video kernels.

File: test/prototype_transforms_kernel_infos.py
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:135]: I intentionally only added a single `KernelInfo` for videos here, since we already have quite a few tests and videos will add even more. Thus, I'll go for 3. of #6622 first before adding more tests here. As explained above, the video kernels are a trivial wrapper around the image kernels for now, so unless I made a mistake while implementing the signature, there should be no new bugs.

File: torchvision/prototype/features/_video.py
Comment (datumbox) [reply torchvision/prototype/features/_video.py:29]: I think for videos we can safely require exactly 4 dimensions.

File: torchvision/prototype/features/_video.py
Comment (datumbox) [reply torchvision/prototype/features/_video.py:56]: Could you add a TODO?

File: torchvision/prototype/features/_video.py
Comment (datumbox) [inline torchvision/prototype/features/_video.py:80]: Why is Video capitalized?

File: torchvision/prototype/transforms/functional/_augment.py
Comment (datumbox) [reply torchvision/prototype/transforms/functional/_augment.py:20]: For those that dont support arbitrary dimensions, you are able to reshape the T dimension into N, send to the Image kernel and then reshape back as a temporarily workaround.

File: torchvision/prototype/transforms/functional/_augment.py
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_augment.py:27]: The proper type for of  `InputType` is defined as `Union[torch.Tensor, PIL.Image.Image, _Feature]`. Note that we include the base class for `_Feature` here not `Image`. So here we imply that BBoxes can be transformed (though I understand you use the JIT-friendly version but things go hand in hand).

One potential approach here is to add `Video` in the `ImageType`  and rename accordingly.

File: torchvision/prototype/transforms/functional/_color.py
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_color.py:26]: Following the pattern with other kernels, shouldn't this be something like:
```suggestion
def adjust_saturation_video(video: features.VideoTypeJIT, saturation_factor: float) -> features.VideoTypeJIT:
```

File: torchvision/prototype/transforms/functional/_misc.py
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_misc.py:17]: All these types such as `features.TensorImageTypeJIT` need to be named to make sense.

File: torchvision/prototype/transforms/functional/_misc.py
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_misc.py:22]: There are so many place that we need to phase out the word image. I've been complaining about this for a very long time 😄  Time to change them.

**Review by datumbox — COMMENTED** [review]
I saw you put comments for us, so from that I assume you want us to have a look. If not the case, feel free to ignore.

File: torchvision/prototype/features/_video.py
Comment (pmeier) [reply torchvision/prototype/features/_video.py:29]: As in a video will always have shape `(F, C, H, W)` rather than arbitrary leading batch dimensions, i.e. `(*, F, C, H, W)`? If yes, what is the reason for this restriction given that all other features don't have this?

Or do you mean we can require the input of the constructor to always have *at least* 4 dimensions?

File: torchvision/prototype/transforms/functional/_augment.py
Comment (pmeier) [reply torchvision/prototype/transforms/functional/_augment.py:20]: I've added #6670. How urgent is it for this to work for videos but not for images? If not as urgent, I would fix the image kernels and benefit from that here.

File: torchvision/prototype/transforms/functional/_augment.py
Comment (pmeier) [reply torchvision/prototype/transforms/functional/_augment.py:27]: What do you think would be a good name? `ImageOrVideoType`?

File: torchvision/prototype/features/_video.py
Comment (datumbox) [reply torchvision/prototype/features/_video.py:29]: I was suggesting the latter. Require the input of the constructor to always have at least 4 dimensions.

File: torchvision/prototype/transforms/functional/_augment.py
Comment (datumbox) [reply torchvision/prototype/transforms/functional/_augment.py:20]: Not urgent, we can do it on a follow up.

File: torchvision/prototype/transforms/functional/_augment.py
Comment (datumbox) [reply torchvision/prototype/transforms/functional/_augment.py:27]: I can live with that for now. We will rethink names before releasing.

File: torchvision/prototype/features/_video.py
Comment (pmeier) [reply torchvision/prototype/features/_video.py:80]: Copy-paste mistake. Good catch!

File: test/prototype_common_utils.py
Comment (jdsgomes) [reply test/prototype_common_utils.py:562]: That's correct. We might support YUV in the future as this is something supported by ffmpeg but I don't think we have any use case for the time being

File: test/prototype_common_utils.py
Comment (jdsgomes) [reply test/prototype_common_utils.py:565]: For now we only support uint8, what are the implications of having float32 in this tuple?

File: test/prototype_common_utils.py
Comment (pmeier) [reply test/prototype_common_utils.py:562]: This also ties into #4029.

File: torchvision/prototype/features/_video.py
Comment (jdsgomes) [reply torchvision/prototype/features/_video.py:64]: Yes that is correct, its the same name used in video clips

File: test/prototype_common_utils.py
Comment (pmeier) [reply test/prototype_common_utils.py:565]: It means all our transforms test that don't explicitly specify the `dtype` will test with videos in both `dtypes`. If we are only supporting `torch.uint8` now, we should remove this here to avoid unnecessary tests that might or might not pass.

Comment (pmeier) [reply test/prototype_common_utils.py:565]:
```suggestion
    dtypes=(torch.uint8,),
```

File: torchvision/prototype/features/_video.py
Comment (YosuaMichael) [reply torchvision/prototype/features/_video.py:29]: @pmeier just curious what `F` stand for in `(F, C, H, W)` ? From what I know, usually the shape is either channel last, `(T, H, W, C)`, or channel first, `(C, T, H, W)`. Where `T` is `temporal` dimension.

Comment (datumbox) [reply torchvision/prototype/features/_video.py:29]: `F` here is the temporal dimension. I recommend using `T` in our comms to avoid misunderstandings.

File: test/prototype_common_utils.py
Comment (bjuncek) [reply test/prototype_common_utils.py:562]: I do not think supporting YUV is quite necessary. YUV should only be used as an intermediate colour space whilst decoding

**Review by bjuncek — COMMENTED** [review]
One high level comment:

At the moment, if i understand it correctly, `Video` feature class only deals with seq-of-images data. 
How hard (or easy) would it be to extend this to additional video modalities (say audio)?

File: torchvision/prototype/features/_video.py
Comment (pmeier) [reply torchvision/prototype/features/_video.py:94]: See https://github.com/pytorch/vision/pull/6667#discussion_r984360727.

**Comment by pmeier** [comment]
> At the moment, if i understand it correctly, `Video` feature class only deals with seq-of-images data.
> How hard (or easy) would it be to extend this to additional video modalities (say audio)?

As is, `Video` is a tensor subclass and thus does not support more than one storage. Thus, I don't think it is possible to bundle video and audio into one tensor. There are ["nested" tensors](https://pytorch.org/docs/stable/nested.html), but the are in prototype state for quite some time now. Plus, the tensors ought to have the same dimensionality (not shape), but that will be quite awkward for audio. I'm guessing the actual shape should be something like `(*, K, L)` like we have in our video datasets now:

https://github.com/pytorch/vision/blob/07ae61bf9c21ddd1d5f65d326aa9636849b383ca/torchvision/datasets/kinetics.py#L72-L73

(Noob question: we don't see `T` here, because the audio is sampled differently, read a lot higher, than the video, right?)

Could you be more specific what kind of transformations on video with audio you have in mind?

File: torchvision/prototype/transforms/functional/_color.py
Comment (pmeier) [reply torchvision/prototype/transforms/functional/_color.py:26]: No. The kernels take plain tensors and return them. As such annotating with `torch.Tensor` is correct as this is not a workaround for JIT, but what we actually want.

File: torchvision/prototype/transforms/_augment.py
Comment (datumbox) [inline torchvision/prototype/transforms/_augment.py:95]:
```suggestion
    def _transform(self, inpt: features.ImageOrVideoType, params: Dict[str, Any]) -> features.ImageOrVideoType:
```

File: torchvision/prototype/transforms/_color.py
Comment (datumbox) [inline torchvision/prototype/transforms/_color.py:113]: We need this everywhere I think:
```suggestion
    def _permute_channels(self, inpt: features.ImageOrVideoType, permutation: torch.Tensor) -> features.ImageOrVideoType:
```

File: torchvision/prototype/transforms/functional/_augment.py
Comment (datumbox) [reply torchvision/prototype/transforms/functional/_augment.py:27]: @pmeier You missed this comment. This should be `ImageOrVideoType` I think. Same on the return type.

**Review by datumbox — COMMENTED** [review]
Overall looks good. A few comments below. We also need to adapt the AutoAugment classes.

One concern I have is if there is a specific transform that doesn't support Videos (squashes dimensions for example) or does something unexpected. I'm happy to go overall to the API and do checks for things we need to do or missed. Another thing we need after this PR as clean-up is remove the `image` wording (both from internal method variables and from arguments like `image_size`).

**Comment by bjuncek** [comment]
>As is, Video is a tensor subclass and thus does not support more than one storage. Thus, I don't think it is possible to bundle video and audio into one tensor.

Ah, ok, makes sense. 

> Noob question: we don't see T here, because the audio is sampled differently, read a lot higher, than the video, right?

Yeah. Basically, audio should be `num_channels` (`K`) x `1-D signal`
On average signal is sampled at 48kHZ, so for a second of stereo audio, you'd have 
`2*48000` size tensor. 

>  what kind of transformations on video with audio you have in mind?

Usually the transforms are handled separately (when I was last working in multimodal video field, we'd apply things separately). So for example in the dataset `__getitem__` we'd have:

```
...
video, audio, info = read_video(path)
if v_transform is not None:
    video = v_transform(video)
if a_transform is not None:
    audio = a_transform(audio)
...
```


This all makes sense though. Let's worry about audio (and if we'd even need it) down the line.

File: torchvision/prototype/features/_video.py
Comment (datumbox) [inline torchvision/prototype/features/_video.py:240]: Nit: This name kinda sucks and I don't have any better naming proposal. What if instead of having a single type that unions both, we just use union on the actual method? I suspect that this might be as informative but less weird. This is just an opinion though and a matter of preference, thoughts @pmeier and @vfdev-5?

File: torchvision/prototype/transforms/_auto_augment.py
Comment (datumbox) [inline torchvision/prototype/transforms/_auto_augment.py:40]: Nit?
```suggestion
        inputs = []
```

File: torchvision/prototype/transforms/_auto_augment.py
Comment (datumbox) [inline torchvision/prototype/transforms/_auto_augment.py:481]: Nit?
```suggestion
            image_or_video = F.pil_to_tensor(orig_image_or_video)
```

File: torchvision/prototype/transforms/functional/_augment.py
Comment (datumbox) [reply torchvision/prototype/transforms/functional/_augment.py:27]: @pmeier You keep resolving this comment but it's still unresolved. This type is wrong, because BBoxes are not suppored. It needs to be `ImageOrVideoType`

File: torchvision/prototype/transforms/functional/_color.py
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_color.py:44]: Question: Should we alias it instead? On the pros side, if someone asks for the name of the method, they will get `adjust_contrast_video` which is what we want. On the cons side, we just have an extra unnecessary method call. No strong opinion, just highlighting it.

File: torchvision/prototype/transforms/functional/_meta.py
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_meta.py:225]: This specific `isinstance` check for image or video is correct, but I think we missed it in all other places on the dispatchers. I think we need to add this everywhere.

**Review by datumbox — APPROVED** [review]
LGTM, after addressing some of the comments below. I'll have an other look on the whole API immediately after merging this to ensure we didn't miss anything.

File: torchvision/prototype/transforms/functional/_augment.py
Comment (pmeier) [reply torchvision/prototype/transforms/functional/_augment.py:27]: True, it seems I keep missing it :facepalm:

File: torchvision/prototype/features/_video.py
Comment (vfdev-5) [reply torchvision/prototype/features/_video.py:240]: What's the difference between `VideoType` and `TensorVideoType` ?

Comment (pmeier) [reply torchvision/prototype/features/_video.py:240]: There is none. It is only there for parity with the image types.

File: torchvision/prototype/transforms/_auto_augment.py
Comment (pmeier) [reply torchvision/prototype/transforms/_auto_augment.py:40]: I renamed to `image_or_videos` to be on par with the `image_or_video` variable in the actual transforms. I didn't like `inputs` here, because that would imply `input = self._extra_image_or_video(...)` and that clashes with the name. Plus, `input` is already used as variable in the actual transformation.

File: torchvision/prototype/transforms/functional/_color.py
Comment (pmeier) [reply torchvision/prototype/transforms/functional/_color.py:44]: >  we just have an extra unnecessary method call. 

That will hardly make a difference. I think we can afford an extra kernel for now. In the future we can revisit this. Especially, if we support 2d images everywhere, most mask kernels could also be aliased. The only exceptions that come to mind are the ones with an interpolation mode.

File: torchvision/prototype/transforms/functional/_meta.py
Comment (pmeier) [reply torchvision/prototype/transforms/functional/_meta.py:225]: I went through the whole API again and found a few places where we need to decide if we want video support there:

- classificaton transforms: `MixUp`, `CutMix`
- object detection transforms: `SimpleCopyPaste`, `RandomIoUCrop`
- deprecated transforms: `Grayscale`, `RandomGrayscale`
- outlier transforms: `FiveCrop`, `TenCrop` (we are not supporting anything besides images here so far)
- image only transforms (by name): `ConvertImageDtype`

For everything else this is fixed.

File: test/prototype_transforms_kernel_infos.py
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:159]: I've added them here to properly test video kernels that have the squash / unsquash "wrappers" now and thus do support arbitrary batch dimensions.

**Comment by datumbox** [comment]
We got breakages on some tests when we throw in batches:
```
TypeError: Input image tensor should have 3 or 4 dimensions, but found 5
_ TestSmoke.test_auto_augment[AugMix-torchvision.prototype.features._video.Video-20] _
Traceback (most recent call last):
  File "/home/runner/work/vision/vision/test/test_prototype_transforms.py", line 183, in test_auto_augment
    transform(input)
  File "/opt/hostedtoolcache/Python/3.7.14/x64/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/_auto_augment.py", line 517, in forward
    aug, transform_id, magnitude, interpolation=self.interpolation, fill=self.fill
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/_auto_augment.py", line 133, in _apply_image_or_video_transform
    return F.adjust_sharpness(image, sharpness_factor=1.0 + magnitude)
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/functional/_color.py", line 84, in adjust_sharpness
    return adjust_sharpness_image_tensor(inpt, sharpness_factor=sharpness_factor)
  File "/home/runner/work/vision/vision/torchvision/transforms/functional_tensor.py", line 845, in adjust_sharpness
    return _blend(img, _blurred_degenerate_image(img), sharpness_factor)
  File "/home/runner/work/vision/vision/torchvision/transforms/functional_tensor.py", line 825, in _blurred_degenerate_image
    result_tmp = conv2d(result_tmp, kernel, groups=result_tmp.shape[-3])
RuntimeError: Expected 3D (unbatched) or 4D (batched) input to conv2d, but got input of size: [1, 1, 1, 23, 15]
_ TestSmoke.test_auto_augment[AugMix-torchvision.prototype.features._video.Video-21] _
Traceback (most recent call last):
  File "/home/runner/work/vision/vision/test/test_prototype_transforms.py", line 183, in test_auto_augment
    transform(input)
  File "/opt/hostedtoolcache/Python/3.7.14/x64/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/_auto_augment.py", line 517, in forward
    aug, transform_id, magnitude, interpolation=self.interpolation, fill=self.fill
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/_auto_augment.py", line 133, in _apply_image_or_video_transform
    return F.adjust_sharpness(image, sharpness_factor=1.0 + magnitude)
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/functional/_color.py", line 84, in adjust_sharpness
    return adjust_sharpness_image_tensor(inpt, sharpness_factor=sharpness_factor)
  File "/home/runner/work/vision/vision/torchvision/transforms/functional_tensor.py", line 845, in adjust_sharpness
    return _blend(img, _blurred_degenerate_image(img), sharpness_factor)
  File "/home/runner/work/vision/vision/torchvision/transforms/functional_tensor.py", line 825, in _blurred_degenerate_image
    result_tmp = conv2d(result_tmp, kernel, groups=result_tmp.shape[-3])
RuntimeError: Expected 3D (unbatched) or 4D (batched) input to conv2d, but got input of size: [1, 1, 1, 26, 19]
_ TestSmoke.test_auto_augment[AugMix-torchvision.prototype.features._video.Video-22] _
Traceback (most recent call last):
  File "/home/runner/work/vision/vision/test/test_prototype_transforms.py", line 183, in test_auto_augment
    transform(input)
  File "/opt/hostedtoolcache/Python/3.7.14/x64/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/_auto_augment.py", line 517, in forward
    aug, transform_id, magnitude, interpolation=self.interpolation, fill=self.fill
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/_auto_augment.py", line 133, in _apply_image_or_video_transform
    return F.adjust_sharpness(image, sharpness_factor=1.0 + magnitude)
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/functional/_color.py", line 84, in adjust_sharpness
    return adjust_sharpness_image_tensor(inpt, sharpness_factor=sharpness_factor)
  File "/home/runner/work/vision/vision/torchvision/transforms/functional_tensor.py", line 845, in adjust_sharpness
    return _blend(img, _blurred_degenerate_image(img), sharpness_factor)
  File "/home/runner/work/vision/vision/torchvision/transforms/functional_tensor.py", line 825, in _blurred_degenerate_image
    result_tmp = conv2d(result_tmp, kernel, groups=result_tmp.shape[-3])
RuntimeError: Expected 3D (unbatched) or 4D (batched) input to conv2d, but got input of size: [1, 3, 3, 24, 22]
_ TestSmoke.test_auto_augment[AugMix-torchvision.prototype.features._video.Video-23] _
Traceback (most recent call last):
  File "/home/runner/work/vision/vision/test/test_prototype_transforms.py", line 183, in test_auto_augment
    transform(input)
  File "/opt/hostedtoolcache/Python/3.7.14/x64/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/_auto_augment.py", line 517, in forward
    aug, transform_id, magnitude, interpolation=self.interpolation, fill=self.fill
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/_auto_augment.py", line 141, in _apply_image_or_video_transform
    return F.equalize(image)
  File "/home/runner/work/vision/vision/torchvision/prototype/transforms/functional/_color.py", line 204, in equalize
    return equalize_image_tensor(inpt)
  File "/home/runner/work/vision/vision/torchvision/transforms/functional_tensor.py", line 900, in equalize
    raise TypeError(f"Input image tensor should have 3 or 4 dimensions, but found {img.ndim}")
TypeError: Input image tensor should have 3 or 4 dimensions, but found 5
```

We might need to review the mitigations.

File: torchvision/prototype/transforms/functional/_color.py
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_color.py:199]: Why not?
```suggestion
        return _FT._equalize_single_image(channels).view(shape)
```

File: torchvision/prototype/transforms/functional/_color.py
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_color.py:62]: Since we are copy-pasting code, let's take a moment to rewrite on our side these 2 checks. If you see what they do internally, they are pretty bad generating multiple duplicate calls. Could you just rewrite these 2 assertions here?

**Comment by pmeier** [comment]
The breakages that we observed in `AugMix` were two-fold:

1. There was an error in the video generation for the tests. I thought I had fixed it in f1e2bfa407b721ef2edda888ee0a6dc7bc3ed1ed but I needed d8945e68914cdbaf5fcdc5848e45e68bf406657f as well.
2. The problem with `AugMix` is that we don't transform the extracted `image` directly, but perform some operations on it first:

    https://github.com/pytorch/vision/blob/7eb5d7fcab73afec976907a855d9e63fa31f5579/torchvision/prototype/transforms/_auto_augment.py#L485

  This effectively unwraps the `features.Video` into a `torch.Tensor`. When we call the dispatchers later 

  https://github.com/pytorch/vision/blob/7eb5d7fcab73afec976907a855d9e63fa31f5579/torchvision/prototype/transforms/_auto_augment.py#L514-L516

  the input gets dispatched to the image kernel. In theory, that is no issue since the video kernel added by this PR does the same. However, as detailed in #6670, `equalize_image_tensor` and `adjust_sharpness_image_tensor` only support 4d inputs. Batched videos are 5d and thus the call fails.

  Previously, we opted to only fix the video kernels in 0d2ad963e27616f53fc9ee24456e43c927b561da. To use this fix, we would need to wrap the inputs to the AA dispatchers into a `Video` again, which might entail a significant runtime cost until #6681 or #6718 are merged. Thus, I've opted to fix the image kernels here since we need to do this anyway at some point.

File: torchvision/prototype/transforms/functional/_color.py
Comment (pmeier) [reply torchvision/prototype/transforms/functional/_color.py:199]: I wanted to avoid that someone copy-pastes `_equalize_single_image` as it can easily be inlined. I'll add a comment and use it here.

File: torchvision/prototype/transforms/functional/_color.py
Comment (pmeier) [reply torchvision/prototype/transforms/functional/_color.py:62]: Sure.

**Review by datumbox — APPROVED** [review]
LGTM, feel free to merge on green CI.

**Pull request merged**
